sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -95,17 +96,20 @@ private Bounded(@Nullable String name, BoundedSource<T> source) {
     }
 
     @Override
-    protected Coder<T> getDefaultOutputCoder() {
+    protected Coder<T> getDefaultOutputCoder() throws CannotProvideCoderException {
       return source.getDefaultOutputCoder();
     }
 
     @Override
     public final PCollection<T> expand(PBegin input) {
       source.validate();
-
-      return PCollection.<T>createPrimitiveOutputInternal(input.getPipeline(),
-          WindowingStrategy.globalDefault(), IsBounded.BOUNDED)
-          .setCoder(getDefaultOutputCoder());
+      try {
+        return PCollection.<T>createPrimitiveOutputInternal(input.getPipeline(),
+            WindowingStrategy.globalDefault(), IsBounded.BOUNDED)
+            .setCoder(getDefaultOutputCoder());
+      } catch (CannotProvideCoderException e) {
```

Comment:
I don't think we should rethrow the exception here. It only means that the source wasn't able to provide a coder for us - so we should just not call .setCoder() ourselves, and the user will need to explicitly .setCoder() on the returned PCollection.

Actually I'm not sure why there even was a .setCoder() call in the original code - it is not present in the Read.Unbounded.expand() method. I think we should just mimic that method and remove this call.